### PR TITLE
Bot templating

### PR DESCRIPTION
**Pull request by scm2217**
Pathfinding and global knowledge. Discussion should occur on how libraries and singletons are setup.

### CODE REVIEW

File: bots/lib/KnowledgeBase/RoomMap.ts [hunk 1]
```diff
@@ -11,51 +11,120 @@ export class RoomMap {
...
         });
       }
 
-      setDifference = new Set([...this.nodes].filter (x => !visited.has(x)));
+      setDifference = new Set([...this.nodes].filter(x => !visited.has(x)));
     }
 
-    return Array.from (startPoints);
+    return Array.from(startPoints);
   }
```
Comment (mitchkm) [inline bots/lib/KnowledgeBase/RoomMap.ts]: In light of potentially making additional graphs of other things, can we pull out/get a graph library for common graph algorithms, following the Set of Nodes, Set of Connections structure.

File: bots/lib/KnowledgeBase/KnowledgeBase.ts [hunk 2]
```diff
@@ -8,4 +12,36 @@ export class KnowledgeBase {
...
+                break;
+              }
+              i++;
+            }
+          }
+        });
+      }
+    }
+    return location;
+  }
```
[the remark this answers is not in the document]
Comment (scm2217) [reply bots/lib/KnowledgeBase/KnowledgeBase.ts]: agree, should look for an alternative way to do this

**Comment by scm2217** [comment]
https://github.com/panoptyk/example-game/blob/db3b8b508fb8bcfef28361fe127a57934e3cbccd/bots/lib/KnowledgeBase/KnowledgeBase.ts#L9-L14

Make this singleton protected instead of private so that KnowledgeBase can be properly extended.

**Comment by mitchkm** [comment]
> https://github.com/panoptyk/example-game/blob/db3b8b508fb8bcfef28361fe127a57934e3cbccd/bots/lib/KnowledgeBase/KnowledgeBase.ts#L7
> 
> Make this singleton protected instead of private so that KnowledgeBase can be properly extended.

I advised not to. Protected is a really bad thing to use in Typescript I feel since it doesn't do anything. (well none of them actually do). That's why I think we should move KnowledgeBase to be a namespace and add everything as such (merging the namespace).

Because the public call can be overwritten it's fine to just have a different variable name for the extended class instance. `_instanceChild` This way both versions are functioning as singletons

**Pull request merged**